Adding metamorpheus

acesnik opened the pull request:
### Adding MetaMorpheus, a software for searching mass spectrometry proteomics data.

---

<!--
Thank you very much for putting in this recipe PR!

This repository is very active, so if you need help with
a PR or once it's ready for review, please let the right people know.
There are language-specific teams for reviewing recipes.

Currently available teams are:
- python `@conda-forge/help-python`
- python/c hybrid `@conda-forge/help-python-c`
- r `@conda-forge/help-r`
- java `@conda-forge/help-java`
- nodejs `@conda-forge/help-nodejs`
- c/c++ `@conda-forge/help-c-cpp`
- perl `@conda-forge/help-perl`
- Julia `@conda-forge/help-julia`
- ruby `@conda-forge/help-ruby`

If your PR doesn't fall into those categories please contact
the full review team `@conda-forge/staged-recipes`.

Due to GitHub limitations first time contributors to conda-forge are unable
to ping these teams. You can [ping the team][1] using a special command in
a comment on the PR to get the attention of the `staged-recipes` team. You can
also consider asking on our [Gitter channel][2] if your recipe isn't reviewed promptly.

[1]: https://conda-forge.org/docs/maintainer/infrastructure.html#conda-forge-admin-please-ping-team
[2]: https://gitter.im/conda-forge/conda-forge.github.io
-->

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/metamorpheus```) and found some lint.

Here's what I've got...


For **recipes/metamorpheus**:

 * Please do not delete the example recipe found in `recipes/example/meta.yaml`.
 * The extra section contained an unexpected subsection name. identifiers is not a valid subsection name.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/metamorpheus```) and found some lint.

Here's what I've got...


For **recipes/metamorpheus**:

 * The extra section contained an unexpected subsection name. identifiers is not a valid subsection name.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/metamorpheus```) and found it was in an excellent condition.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/metamorpheus```) and found some lint.

Here's what I've got...


For **recipes/metamorpheus**:

 * Failed to even lint the recipe, probably because of a conda-smithy bug :cry:. This likely indicates a problem in your `meta.yaml`, though. To get a traceback to help figure out what's going on, install conda-smithy and run `conda smithy recipe-lint .` from the recipe directory.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/metamorpheus```) and found some lint.

Here's what I've got...


For **recipes/metamorpheus**:

 * Selectors are suggested to take a ``<two spaces>#<one space>[<expression>]`` form. See lines [24]

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/metamorpheus```) and found it was in an excellent condition.

acesnik commented:
@conda-forge-admin, please restart ci

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/metamorpheus```) and found some lint.

Here's what I've got...


For **recipes/metamorpheus**:

 * `noarch` packages can't have selectors. If the selectors are necessary, please remove `noarch: generic`.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/metamorpheus```) and found it was in an excellent condition.

acesnik commented:
@bgruening @ZimmerA @dhirschfeld 

Hi all, this PR is ready to review when you have the chance. It aims to add a feedstock for MetaMorpheus, a program for searching mass spectrometry proteomics data, a type of biological data. 

I've chosen to make this PR to conda-forge, rather than bioconda, because conda-forge allows Windows builds and bioconda does not, and making this software available for Windows via conda is a priority for us.

Thanks! -AC

dopplershift commented on .azure-pipelines/azure-pipelines-linux.yml:
```diff
@@ -1,70 +1,5 @@
 jobs:
-- job: linux_64
```
This PR shouldn't be changing files outside the recipe directory.

acesnik replied:
I rolled these changes back.

dopplershift reviewed (changes requested):
I'm not an expert on the C-building side of things, but at a minimum you'll need to rollback the changes to files outside your recipe.

dopplershift commented on conda-forge.yml:
```diff
@@ -1 +1,3 @@
 channel_priority: strict
+os_version:
```
Same here.

acesnik commented:
@dopplershift Thanks for taking a look! Let me see if I can explain those changes.

This PR only works on cos7 systems because it has `dotnet` (https://github.com/conda-forge/dotnet-feedstock) as a dependency, which currently is only built to run on cos7 systems. I changed these files outside the recipe because this PR shouldn't be built on other linux systems for that reason. It's noted in the knowledgebase that the change to `conda-forge.yaml` is the way to do this (https://conda-forge.org/docs/maintainer/knowledge_base.html#using-centos-7), which changes how `conda smithy` generates the code for these builds when a feedstock is updated, but since `conda smithy` isn't used here in the staged recipes, I changed that build information manually.

dhirschfeld commented:
If you make changes here it will affect *all* future recipes so I believe the policy is to not make changes to the CI in staged recipes but instead to simply skip the problematic builds and enable on the feedstock after creation.

acesnik commented:
Thanks for explaining how to handle the CI here. I've rolled back the changes outside the recipe directory and skipped the linux build for now.

dhirschfeld commented on recipes/metamorpheus/meta.yaml:
```diff
@@ -0,0 +1,37 @@
+    - {{ compiler('c') }}
+    - sysroot_linux-64 2.17  # [linux64]
+  run:
+    - dotnet =3.1
```
It doesn't work with 5?

acesnik replied:
Not yet, but that's in the works. https://github.com/smith-chem-wisc/MetaMorpheus/pull/2047

acesnik replied:
I'll try removing it and see what happens and revert if needed. I think I remember that if the NetCore.App 3.1 runtime isn't present, it doesn't succeed.

acesnik replied:
This is what I see with 5.0:

![image](https://user-images.githubusercontent.com/16342951/113702665-1daf5980-96da-11eb-9b5a-0f29569f80f3.png)

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=300108&view=logs&jobId=240f1fee-52bc-5498-a14a-8361bde76ba0&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=23010c01-6a9d-5f94-9037-792618644b77

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipes/mplhep_data```, ```recipes/ncagg```) and found some lint.

Here's what I've got...


For **recipes/mplhep_data**:

 * Feedstock with the same name exists in conda-forge.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/metamorpheus```) and found it was in an excellent condition.

dhirschfeld commented on recipes/metamorpheus/build.sh:
```diff
@@ -0,0 +1,15 @@
+mkdir -p $PREFIX/bin
+cp -r $SRC_DIR/* $PREFIX
+
+if [[ "${build_platform}" == "win-64" ]]; then
+    mv $PREFIX/CMD.exe $PREFIX/metamorpheus.exe
+else
+    cp $RECIPE_DIR/metamorpheus $PREFIX/metamorpheus
```
Normally binaries are installed into `$PREFIX/bin` rather than directly into `$PREFIX`

dhirschfeld replied:
It looks like all the `metamorpheus` files are being installed into `$PREFIX`!

![image](https://user-images.githubusercontent.com/881019/113706821-760a4980-9722-11eb-801d-66a36b4bc2ab.png)

* https://artprodeus21.artifacts.visualstudio.com/A910fa339-c7c2-46e8-a579-7ea247548706/84710dde-1620-425b-80d0-4cf5baca359d/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2NvbmRhLWZvcmdlL3Byb2plY3RJZC84NDcxMGRkZS0xNjIwLTQyNWItODBkMC00Y2Y1YmFjYTM1OWQvYnVpbGRJZC8zMDAxMTIvYXJ0aWZhY3ROYW1lL2NvbmRhX3BrZ3Nfd2lu0/content?format=zip

dhirschfeld replied:
I think `metamorpheus` should probably be installed into it's own folder - maybe `$PREFIX/dotnet/tools/metamorpheus`? That's the location I installed [`dotnet-interactive`](https://github.com/conda-forge/dotnet-interactive-feedstock/blob/master/recipe/bld.bat)

To make the executable available on the path you could then add an executable shim in `$PREFIX/bin` or maybe `$PREFIX/dotnet/tools`
e.g. https://github.com/conda-forge/dotnet-feedstock/blob/master/recipe/dotnet.cmd

dhirschfeld commented on recipes/metamorpheus/build.sh:
```diff
@@ -3,13 +3,13 @@ set -eox pipefail
 
-mkdir -p $PREFIX/bin
-cp -r $SRC_DIR/* $PREFIX
+mkdir -p $PREFIX/bin/tools
```
`tools` is a subfolder of `$PREFIX/dotnet` rather than `$PREFIX/bin`

dhirschfeld replied:
With dotnet 5 `DOTNET_ROOT` defines the folder where `dotnet` is installed (it's different on unix/win)

acesnik replied:
Got it; thanks for the help. Just pushed that change.

dhirschfeld commented on recipes/metamorpheus/build.sh:
```diff
@@ -3,13 +3,13 @@ set -eox pipefail
-mkdir -p $PREFIX/bin
-cp -r $SRC_DIR/* $PREFIX
+mkdir -p $PREFIX/bin/tools
+cp -r $SRC_DIR/* $PREFIX/bin
```
This just copies all your application files into the `bin` directory which doesn't seem like a great idea - it's not well encapsulated and will inevitably lead to files being clobbered. Instead, I think your application should be installed into it's own folder (somewhere) rather than into a common folder. I'd suggest `$DOTNET_ROOT/tools/metamorpheus`

acesnik replied:
Thanks for all the advice. I'm still learning how these end up in relation to other packages.

This makes sense with your suggested change below regarding where the DLL should be. I'll go ahead and make those changes.

dhirschfeld commented on recipes/metamorpheus/metamorpheus:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+
+DIR="$(cd "$(dirname "$(dirname "$0")")" && pwd)"
+dotnet $DIRDIR/CMD.dll $@
```
```suggestion
dotnet $DOTNET_ROOT/tools/metamorpheus/CMD.dll $@
```

dhirschfeld commented on recipes/metamorpheus/build.sh:
```diff
@@ -3,13 +3,13 @@ set -eox pipefail
 cp -r $SRC_DIR/* $PREFIX/bin
 
 if [[ "${build_platform}" == "win-64" ]]; then
     mv $PREFIX/bin/CMD.exe $PREFIX/bin/metamorpheus.exe
```
Does this actually work? I thought the exe might expect to find it's associated dlls alongside it.

dhirschfeld commented on recipes/metamorpheus/build.sh:
```diff
@@ -0,0 +1,16 @@
+cp -r $SRC_DIR/* $METAMORPHEUS_ROOT
+
+if [[ "${build_platform}" == "win-64" ]]; then
+    mv $METAMORPHEUS_ROOT/CMD.exe $METAMORPHEUS_ROOT/metamorpheus.exe
```
```suggestion
    mv $METAMORPHEUS_ROOT/CMD.exe $METAMORPHEUS_ROOT/metamorpheus.exe
    cp $RECIPE_DIR/metamorpheus.cmd $DOTNET_ROOT/tools/
```
**`metamorpheus.cmd:`**
```batch
@"%~dp0\metamorpheus\metamorpheus.exe" %*
```

dhirschfeld commented:
I thought that would work. There's nothing *obvious* that I can see wrong.

Unfortunately since the test failed no artifact was published:
https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=300181&view=artifacts&type=publishedArtifacts

In situations like this I often make the test a no-op so I can then check the package published as an artifact.

acesnik commented:
Okay, with making the `$PREFIX/bin` folder and setting `DOTNET_ROOT` before building and packaging, it looks like it's in better shape!

dhirschfeld commented:
Package structure looks good:

<details>

```
CONDA_ENV_ROOT
│
├───bin
│       metamorpheus.cmd
│
├───dotnet
│   └───tools
│       └───metamorpheus
│           │   BayesianEstimation.dll
│           │   build.sh
│           │   build_env_setup.bat
│           │   Chemistry.dll
│           │   CMD.deps.json
│           │   CMD.dll
│           │   CMD.dll.config
│           │   CMD.exe
│           │   CMD.pdb
│           │   CMD.runtimeconfig.dev.json
│           │   CMD.runtimeconfig.json
│           │   CommandLine.dll
│           │   conda_build.bat
│           │   Easy.Common.dll
│           │   EngineLayer.dll
│           │   EngineLayer.pdb
│           │   FlashLFQ.dll
│           │   MassSpectrometry.dll
│           │   MathNet.Numerics.dll
│           │   metadata_conda_debug.yaml
│           │   Mgf.dll
│           │   Microsoft.ML.Core.dll
│           │   Microsoft.ML.CpuMath.dll
│           │   Microsoft.ML.Data.dll
│           │   Microsoft.ML.DataView.dll
│           │   Microsoft.ML.FastTree.dll
│           │   Microsoft.ML.KMeansClustering.dll
│           │   Microsoft.ML.PCA.dll
│           │   Microsoft.ML.StandardTrainers.dll
│           │   Microsoft.ML.Transforms.dll
│           │   MzIdentML.dll
│           │   MzLibUtil.dll
│           │   MzML.dll
│           │   mzPlot.dll
│           │   NetSerializer.dll
│           │   Nett.dll
│           │   Newtonsoft.Json.dll
│           │   OpenMcdf.dll
│           │   OpenMcdf.Extensions.dll
│           │   PepXML.dll
│           │   Proteomics.dll
│           │   SharpLearning.Common.Interfaces.dll
│           │   SharpLearning.Containers.dll
│           │   SharpLearning.DecisionTrees.dll
│           │   SharpLearning.InputOutput.dll
│           │   SharpLearning.Optimization.dll
│           │   SharpLearning.RandomForest.dll
│           │   System.CodeDom.dll
│           │   TaskLayer.dll
│           │   TaskLayer.pdb
│           │   ThermoFisher.CommonCore.BackgroundSubtraction.dll
│           │   ThermoFisher.CommonCore.Data.dll
│           │   ThermoFisher.CommonCore.MassPrecisionEstimator.dll
│           │   ThermoFisher.CommonCore.RawFileReader.dll
│           │   ThermoRawFileReader.dll
│           │   UsefulProteomicsDatabases.dll
│           │
│           ├───Contaminants
│           │       MetaMorpheusContaminants.xml
│           │
│           ├───Data
│           │       combos.txt
│           │       Crosslinkers.tsv
│           │       elements.dat
│           │       PSI-MOD.obo.xml
│           │       ptmlist.txt
│           │       SmallCalibratible_Yeast.mzML
│           │       SmallYeast.fasta
│           │       unimod.xml
│           │
│           ├───Glycan_Mods
│           │   ├───NGlycan
│           │   │       NGlycan182.gdb
│           │   │
│           │   └───OGlycan
│           │           OGlycan.gdb
│           │           Olgycan Database 28 glycans plus Sodium 56 total.txt
│           │           Olgycan Database 28 glycans plus Sulfo 56 total.txt
│           │           Olgycan Database 28 glycans.txt
│           │           Olgycan Database 32 glycans.txt
│           │           Olgycan Database 36 glycans with Mann.txt
│           │
│           ├───Mods
│           │       aListOfmods.txt
│           │       glyco.txt
│           │       Mods.txt
│           │       ProteaseMods.txt
│           │       substitutions.txt
│           │       surfactants.txt
│           │       tmt.txt
│           │
│           ├───ProteolyticDigestion
│           │       proteases.tsv
│           │
│           └───runtimes
│               ├───linux-x64
│               │   └───native
│               │           libCpuMathNative.so
│               │           libFastTreeNative.so
│               │           libLdaNative.so
│               │
│               ├───osx-x64
│               │   └───native
│               │           libCpuMathNative.dylib
│               │           libFastTreeNative.dylib
│               │           libLdaNative.dylib
│               │
│               ├───win-x64
│               │   └───native
│               │           CpuMathNative.dll
│               │           FastTreeNative.dll
│               │           LdaNative.dll
│               │
│               └───win-x86
│                   └───native
│                           CpuMathNative.dll
│                           FastTreeNative.dll
│                           LdaNative.dll
│
└───info
    │   about.json
    │   files
    │   git
    │   hash_input.json
    │   has_prefix
    │   index.json
    │   paths.json
    │
    ├───licenses
    │       LICENSE.txt
    │
    ├───recipe
    │       bld.bat
    │       build.sh
    │       conda_build_config.yaml
    │       LICENSE.txt
    │       meta.yaml
    │       meta.yaml.template
    │       metamorpheus
    │       metamorpheus.cmd
    │
    └───test
            run_test.bat
```
</details>

dhirschfeld reviewed (approved):
LGTM

dhirschfeld commented:
LGTM however I can't merge so someone from @conda-forge/staged-recipes will have to weigh in...

acesnik commented:
@dopplershift Do you have permissions to merge this request? In addition to some other improvements, I've removed the changes from outside the `recipe` directory as you requested above.

-AC

acesnik commented:
Made a clean version of this PR, so it's easier for others to review. https://github.com/conda-forge/staged-recipes/pull/14623

bernt-matthias commented:
Looks good to me

acesnik commented:
Merged here https://github.com/conda-forge/staged-recipes/pull/14623
